### PR TITLE
[WASMFS] Use vector instead of map. NFC

### PR DESCRIPTION
**Pull request by ethanalee**
Relevant Issue: #15041 

To reduce code size, remove `std::map` from `WasmFS`.

### CODE REVIEW

**Comment by kripken** [comment]
How much does code size change?

**Comment by ethanalee** [comment]
> How much does code size change?
Used the `sqlite` benchmark tests.

With Map:
`size:  1084182, compressed:   427473`

Without Map:
`size:  1078670, compressed:   426360 `

**Comment by kripken** [comment]
Interesting, thanks... 5K is a lot just for using one map. In JS that would be free :sunglasses:

**Comment by kripken** [comment]
Please measure the code size change here with #15723 as that might make a difference.

**Comment by ethanalee** [comment]
> Please measure the code size change here with #15723 as that might make a difference.

Here is the new size:

`size:  1052109, compressed:   416076`

**Comment by kripken** [comment]
Sorry, I meant to compare the difference. That is: assuming that PR has landed, what does the size look like with a map, and what does the size look like with a vector?

**Comment by ethanalee** [comment]
> 1052109

Comparison:

With `std::map`:

`size:  1053275, compressed:   416419`

Without `std::map`:

`size: 1052109, compressed: 416076`

**Comment by kripken** [comment]
Thanks!

Ok, down to 1K. Probably still worth it, although I'm less sure.

**Comment by ethanalee** [comment]
> Thanks!
> 
> Ok, down to 1K. Probably still worth it, although I'm less sure.

Mmm, well we do incur a linear look up time now.

File: tests/wasmfs/wasmfs_getdents.out [hunk 1]
```diff
@@ -50,17 +50,17 @@ d.d_off = 280
 d.d_reclen = 280
 d.d_type = directory
 
-d.d_name = stderr
```
Comment (kripken) [inline tests/wasmfs/wasmfs_getdents.out:53]: Was this order not deterministic before?

File: system/lib/wasmfs/file.h [hunk 2]
```diff
@@ -263,6 +222,12 @@ class Directory : public File {
       return {};
     }
   }
+
+protected:
+  std::vector<Entry> entries;
```
Comment (kripken) [inline system/lib/wasmfs/file.h:227]: Let's at least add a comment here saying that a vector saves code size but does lead to linear time in certain operations etc.

File: tests/wasmfs/wasmfs_getdents.out (hunk 1 above)
Comment (ethanalee) [reply tests/wasmfs/wasmfs_getdents.out:53]: It was deterministic, but didn't display in terms of insertion order.

File: system/lib/wasmfs/file.h (hunk 2 above)
Comment (ethanalee) [reply system/lib/wasmfs/file.h:227]: Sounds good!

**Review by kripken — APPROVED** [review]

File: system/lib/wasmfs/file.cpp [hunk 3]
```diff
@@ -32,13 +32,59 @@ void DataFile::Handle::preloadFromJS(int index) {
...
+  // child should be atomic. The state should not be mutated in between.
+  auto unlinked = getEntry(pathName)->locked();
+  unlinked.setParent({});
+
+  for (auto it = getDir()->entries.begin(); it != getDir()->entries.end();
+       it++) {
+    if (it->name == pathName) {
+      getDir()->entries.erase(it);
+      return;
+    }
```
Comment (tlively) [inline system/lib/wasmfs/file.cpp:75]: This loop could be replaced with a `std::remove` + `std::erase` or `std::find` + `vector::erase`.

File: system/lib/wasmfs/file.cpp [hunk 4]
```diff
@@ -32,13 +32,59 @@ void DataFile::Handle::preloadFromJS(int index) {
...
   }
 }
+
+std::string Directory::Handle::getName(std::shared_ptr<File> target) {
+  for (const auto& entry : getDir()->entries) {
+    if (entry.file == target) {
+      return entry.name;
+    }
+  }
```
Comment (tlively) [inline system/lib/wasmfs/file.cpp:84]: This could be a `std::find`

**Review by tlively — APPROVED** [review]

File: system/lib/wasmfs/file.cpp (hunk 4 above)
Comment (ethanalee) [reply system/lib/wasmfs/file.cpp:84]: Might be a little bit more verbose, since we need a lambda comparator.

**Comment by tlively** [comment]
It looks like there are a bunch of changes to the file table handle here as well. Were those meant to be in a separate PR?

**Comment by ethanalee** [comment]
> It looks like there are a bunch of changes to the file table handle here as well. Were those meant to be in a separate PR?

I can split it out. I just thought that since we were refactoring some of the accessor functions for the handle, we might as well split them out into the `.cpp` file as well to clean up the header.

**Review by tlively — APPROVED** [review]

**Pull request merged**